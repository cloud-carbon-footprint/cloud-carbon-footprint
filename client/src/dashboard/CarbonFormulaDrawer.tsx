/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent, useState } from 'react'
import clsx from 'clsx'
import { createStyles, makeStyles } from '@material-ui/core/styles'
import { Close, Info, OpenInNew } from '@material-ui/icons'
import { Typography, Drawer, Divider, IconButton, Link } from '@material-ui/core'

const drawerWidth = 340

const useStyles = makeStyles(({ palette, transitions, spacing, typography, breakpoints, mixins }) => {
  return createStyles({
    infoButton: {
      margin: spacing(10, 'auto'),
      color: palette.type === 'dark' ? palette.text.primary : palette.primary.main,
    },
    closeButtonContainer: {
      textAlign: 'right',
    },
    drawer: {
      width: drawerWidth,
      flexShrink: 0,
      whiteSpace: 'nowrap',
    },
    drawerOpen: {
      width: drawerWidth,
      transition: transitions.create('width', {
        easing: transitions.easing.sharp,
        duration: transitions.duration.enteringScreen,
      }),
    },
    drawerClose: {
      transition: transitions.create('width', {
        easing: transitions.easing.sharp,
        duration: transitions.duration.leavingScreen,
      }),
      overflowX: 'hidden',
      width: spacing(4) + 1,
      [breakpoints.up('sm')]: {
        width: spacing(6) + 1,
      },
    },
    toolbar: {
      padding: spacing(10, 1),
      // necessary for content to be below app bar
      ...mixins.toolbar,
    },
    hide: {
      display: 'none',
    },
    contentTitle: {
      padding: spacing(2),
      fontWeight: typography.fontWeightBold,
    },
    contentBold: {
      padding: spacing(2, 2, 0, 2),
      fontWeight: typography.fontWeightBold,
    },
    content: {
      padding: spacing(2),
      whiteSpace: 'pre-line',
      fontSize: typography.body2.fontSize,
    },
    methodology: {
      padding: spacing(2),
      display: 'flex',
      color: palette.extLink,
    },
    openIcon: {
      marginLeft: '8px',
    },
  })
})

export const CarbonFormulaDrawer: FunctionComponent = () => {
  const classes = useStyles()
  const [open, setOpen] = useState(false)

  const handleDrawerOpen = () => {
    setOpen(true)
  }

  const handleDrawerClose = () => {
    setOpen(false)
  }

  return (
    <Drawer
      anchor="right"
      variant="permanent"
      className={clsx(classes.drawer, {
        [classes.drawerOpen]: open,
        [classes.drawerClose]: !open,
      })}
      classes={{
        paper: clsx({
          [classes.drawerOpen]: open,
          [classes.drawerClose]: !open,
        }),
      }}
    >
      <IconButton
        color="inherit"
        aria-label="open drawer"
        onClick={handleDrawerOpen}
        edge="start"
        className={clsx(classes.infoButton, {
          [classes.hide]: open,
        })}
      >
        <Info data-testid="infoIcon" />
      </IconButton>
      <div
        className={clsx(classes.toolbar, {
          [classes.hide]: !open,
        })}
      >
        <div className={classes.closeButtonContainer}>
          <IconButton onClick={handleDrawerClose}>
            <Close data-testid="closeIcon" />
          </IconButton>
        </div>
        <Typography className={classes.contentTitle} component="p">
          How do we get our carbon estimates?
        </Typography>
        <Divider />
        <Typography className={classes.content} component="p">
          Our CO2e Estimate Formula:
        </Typography>
        <Typography className={classes.content} component="p">
          (Cloud provider service usage) x (Cloud provider Power Usage Effectiveness [PUE]) x (Cloud energy conversion
          factors [Wh]) x (EPA [US] or carbonfootprint.com [Non-US] grid emissions factors [CO2e])
        </Typography>
        <Divider />
        <Typography className={classes.content} component="p">
          Currently we estimate CO2e emissions for cloud compute and storage services. Networking and memory services
          usage are not estimated yet due to their comparatively small footprint and current lack of available energy
          conversion factors.
        </Typography>
        <Link
          href="https://github.com/ThoughtWorks-Cleantech/cloud-carbon-footprint/blob/trunk/METHODOLOGY.md"
          target="_blank"
          rel="noopener"
          className={classes.methodology}
        >
          Read our full methodology here <OpenInNew fontSize={'small'} className={classes.openIcon}></OpenInNew>
        </Link>
      </div>
    </Drawer>
  )
}
