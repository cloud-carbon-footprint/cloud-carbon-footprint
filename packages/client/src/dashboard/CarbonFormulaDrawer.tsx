/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent, useState } from 'react'
import clsx from 'clsx'
import { createStyles, makeStyles } from '@material-ui/core/styles'
import { Close, Info, OpenInNew } from '@material-ui/icons'
import {
  Typography,
  Drawer,
  Divider,
  IconButton,
  Link,
} from '@material-ui/core'

const drawerWidth = 340

const useStyles = makeStyles(
  ({ palette, transitions, spacing, typography, breakpoints, mixins }) => {
    return createStyles({
      infoButton: {
        margin: spacing(10, 'auto'),
        color:
          palette.type === 'dark' ? palette.text.primary : palette.primary.main,
      },
      closeButtonContainer: {
        textAlign: 'right',
      },
      drawer: {
        width: drawerWidth,
        flexShrink: 0,
        whiteSpace: 'nowrap',
      },
      drawerOpen: {
        width: drawerWidth,
        transition: transitions.create('width', {
          easing: transitions.easing.sharp,
          duration: transitions.duration.enteringScreen,
        }),
      },
      drawerClose: {
        transition: transitions.create('width', {
          easing: transitions.easing.sharp,
          duration: transitions.duration.leavingScreen,
        }),
        overflowX: 'hidden',
        width: spacing(4) + 1,
        [breakpoints.up('sm')]: {
          width: spacing(6) + 1,
        },
      },
      toolbar: {
        padding: spacing(10, 1),
        // necessary for content to be below app bar
        ...mixins.toolbar,
      },
      hide: {
        display: 'none',
      },
      contentTitle: {
        padding: spacing(2),
        fontWeight: typography.fontWeightBold,
      },
      contentBold: {
        padding: spacing(2, 2, 0, 2),
        fontWeight: typography.fontWeightBold,
      },
      content: {
        padding: spacing(2),
        whiteSpace: 'pre-line',
        fontSize: typography.body2.fontSize,
      },
      methodology: {
        padding: spacing(2),
        display: 'flex',
        color: palette.extLink,
      },
      openIcon: {
        marginLeft: '8px',
      },
    })
  },
)

export const CarbonFormulaDrawer: FunctionComponent = () => {
  const classes = useStyles()
  const [open, setOpen] = useState(false)

  const handleDrawerOpen = () => {
    setOpen(true)
  }

  const handleDrawerClose = () => {
    setOpen(false)
  }

  const drawerStatus = open ? 'open' : 'closed'
  return (
    <Drawer
      anchor="right"
      variant="permanent"
      id={`drawer-` + drawerStatus}
      className={clsx(classes.drawer, {
        [classes.drawerOpen]: open,
        [classes.drawerClose]: !open,
      })}
      classes={{
        paper: clsx({
          [classes.drawerOpen]: open,
          [classes.drawerClose]: !open,
        }),
      }}
    >
      <IconButton
        color="inherit"
        aria-label="open drawer"
        onClick={handleDrawerOpen}
        edge="start"
        className={clsx(classes.infoButton, {
          [classes.hide]: open,
        })}
        id="info-button"
      >
        <Info data-testid="infoIcon" />
      </IconButton>
      <div
        className={clsx(classes.toolbar, {
          [classes.hide]: !open,
        })}
      >
        <div
          className={classes.closeButtonContainer}
          id="close-button-container"
        >
          <IconButton onClick={handleDrawerClose}>
            <Close data-testid="closeIcon" />
          </IconButton>
        </div>
        <Typography className={classes.contentTitle} component="p">
          How do we get our carbon estimates?
        </Typography>
        <Divider />
        <Typography className={classes.content} component="p">
          Our CO2e Estimate Formula:
        </Typography>
        <Typography className={classes.content} component="p">
          (Cloud provider service usage) x (Cloud provider Power Usage
          Effectiveness [PUE]) x (Cloud energy conversion factors [kWh]) x (EPA
          [US] or carbonfootprint.com [Non-US] grid emissions factors [CO2e])
        </Typography>
        <Divider />
        <Typography className={classes.content} component="p">
          Currently we estimate CO2e emissions for cloud compute, storage and
          networking. Memory usage is not estimated yet due to their arguably
          comparatively small footprint and current lack of available energy
          conversion factors.
          <br />
          <br />
          Emissions data points marked with an * have been estimated with
          average CPU Utilization because the actual CPU Utilization is not
          available.
        </Typography>
        <Link
          href="https://github.com/ThoughtWorks-Cleantech/cloud-carbon-footprint/blob/trunk/microsite/docs/Methodology.md"
          target="_blank"
          rel="noopener"
          className={classes.methodology}
        >
          Read more in our full methodology here{' '}
          <OpenInNew
            fontSize={'small'}
            className={classes.openIcon}
          ></OpenInNew>
        </Link>
      </div>
    </Drawer>
  )
}
