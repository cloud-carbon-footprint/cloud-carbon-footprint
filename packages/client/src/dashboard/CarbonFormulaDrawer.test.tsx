/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React from 'react'
import { create } from 'react-test-renderer'
import { CarbonFormulaDrawer } from './CarbonFormulaDrawer'
import { fireEvent, render } from '@testing-library/react'

import { act } from 'react-dom/test-utils'

describe('CarbonFormulaDrawer', () => {
  it('renders', () => {
    const root = create(<CarbonFormulaDrawer />)
    expect(root.toJSON()).toMatchSnapshot()
  })

  it('opens and closes drawer correctly', async () => {
    const { getByText, getByTestId } = render(<CarbonFormulaDrawer />)
    const infoIcon = getByTestId('infoIcon')
    const carbonFormulaTitle = getByText('How do we get our carbon estimates?')

    expect(carbonFormulaTitle).not.toBeVisible()
    expect(infoIcon).toBeInstanceOf(SVGSVGElement)

    act(() => {
      fireEvent.click(infoIcon)
    })

    const closeIcon = getByTestId('closeIcon')
    expect(closeIcon).toBeInstanceOf(SVGSVGElement)
    expect(carbonFormulaTitle).toBeVisible()

    act(() => {
      fireEvent.click(closeIcon)
    })

    expect(carbonFormulaTitle).not.toBeVisible()
    expect(infoIcon).toBeVisible()
  })
})
