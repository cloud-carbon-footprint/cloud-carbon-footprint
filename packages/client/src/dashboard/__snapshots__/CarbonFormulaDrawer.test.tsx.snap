// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarbonFormulaDrawer renders 1`] = `
<div
  className="MuiDrawer-root MuiDrawer-docked makeStyles-drawer-3 makeStyles-drawerClose-5"
  id="drawer-closed"
>
  <div
    className="MuiPaper-root MuiDrawer-paper makeStyles-drawerClose-5 MuiDrawer-paperAnchorRight MuiDrawer-paperAnchorDockedRight MuiPaper-elevation0"
  >
    <button
      aria-label="open drawer"
      className="MuiButtonBase-root MuiIconButton-root makeStyles-infoButton-1 MuiIconButton-colorInherit MuiIconButton-edgeStart"
      disabled={false}
      id="info-button"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          data-testid="infoIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
          />
        </svg>
      </span>
    </button>
    <div
      className="makeStyles-toolbar-6 makeStyles-hide-7"
    >
      <div
        className="makeStyles-closeButtonContainer-2"
        id="close-button-container"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              data-testid="closeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
      </div>
      <p
        className="MuiTypography-root makeStyles-contentTitle-8 MuiTypography-body1"
      >
        How do we get our carbon estimates?
      </p>
      <hr
        className="MuiDivider-root"
      />
      <p
        className="MuiTypography-root makeStyles-content-10 MuiTypography-body1"
      >
        Our CO2e Estimate Formula:
      </p>
      <p
        className="MuiTypography-root makeStyles-content-10 MuiTypography-body1"
      >
        (Cloud provider service usage) x (Cloud provider Power Usage Effectiveness [PUE]) x (Cloud energy conversion factors [kWh]) x (EPA [US] or carbonfootprint.com [Non-US] grid emissions factors [CO2e])
      </p>
      <hr
        className="MuiDivider-root"
      />
      <p
        className="MuiTypography-root makeStyles-content-10 MuiTypography-body1"
      >
        Currently we estimate CO2e emissions for cloud compute, storage and networking. Memory usage is not estimated yet due to their arguably comparatively small footprint and current lack of available energy conversion factors.
        <br />
        <br />
        Emissions data points marked with an * have been estimated with average CPU Utilization because the actual CPU Utilization is not available.
      </p>
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover makeStyles-methodology-11 MuiTypography-colorPrimary"
        href="https://github.com/ThoughtWorks-Cleantech/cloud-carbon-footprint/blob/trunk/METHODOLOGY.md"
        onBlur={[Function]}
        onFocus={[Function]}
        rel="noopener"
        target="_blank"
      >
        Read more in our full methodology here
         
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root makeStyles-openIcon-12 MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
          />
        </svg>
      </a>
    </div>
  </div>
</div>
`;
